peps/pep-0780.rst

```diff
@@ -0,0 +1,447 @@
+that the GIL-enabled Python will be completely phased out (it may remain
+available as a non standard build option). If it persists, the inverse need for
+the ABI feature detection may arise.
+
+Indeed, in all three phases it may be necessary for package authors to choose
+specific versions of their dependencies based on the ABI features, with a shift
+from GIL-enabled as default to free-threading as default over time.
+
+The ABI features are designed with this in mind to guarantee usefulness and
+simplicity for the foreseeable future in a changing Python ecosystem.
+
+Specification
+=============
+
+The keywords "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**",
+"**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**",
+and "**OPTIONAL**"" in this document are to be interpreted as described in
+:rfc:`RFC 2119 <2119>`.
+
+Concepts
+--------
+
+ABI features are a clear description of certain properties of the Python
+interpreter. While some of these features can be queried already today, they
+are not easily nor uniformly accessible.
+
+Hence, the proposed feature MUST be made available via the Python standard
+library as ``sys.abi_features``, and as the new environment marker variable
+``sys_abi_features``.
+
+ABI Features
+''''''''''''
+
+ABI features are intrinsic properties of the Python interpreter, expressed as
+simple, understandable strings.
+However, not all features are equally applicable to all Python interpreters or
+Python versions. For example, the distinction between free-threaded and
+GIL-enabled interpreters is only relevant for CPython 3.13 onwards, but the bitness
+of the interpreter is relevant for all interpreters.
+
+All interpreters MUST handle the following ABI features as stated.
+ABI features that are restricted to particular interpreters MUST NOT be
+provided by other interpreters.
+The features are subdivided into groups and for each group there MUST be
+exactly one feature present, except when the group is marked as optional, in
+which case there MUST be at most one feature present.
+
+``free-threading`` or ``gil-enabled`` (only CPython)
+    If the Python interpreter is free-threaded, the ``free-threading`` feature
+    MUST be present and the ``gil-enabled`` feature MUST NOT be present.
+    Otherwise, the ``gil-enabled`` feature MUST be present and the
+    ``free-threading`` feature MUST NOT be present.
+
+``debug`` (only CPython, optional)
+    This ABI feature is reserved for the ``--with-pydebug`` build of CPython.
+    If the interpreter is a CPython interpreter with ``Py_DEBUG`` capabilities,
+    the ``debug`` feature MUST be present.
+    On POSIX systems, this corresponds to ``"d" in sys.abiflags``.
+
+``32-bit`` or ``64-bit`` (optional)
+    The bitness of the interpreter, that is, whether it is a 32-bit or 64-bit
+    build [#bitness]_. If the bitness is unknown or neither 32-bit nor 64-bit,
+    this feature MUST NOT be present.
```

Comment:
I'm unclear on the rationale for having bitness be optional, especially given the motivating example of replacing `sys.maxsize`.

Do these values work for e.g. MicroPython?

Reply:
We are not aware of other needed values, certainly not in CPython. MicroPython is an interesting case, as it [supports some 16-bit platforms](https://github.com/micropython/micropython?tab=readme-ov-file#supported-platforms--architectures). It's reasonable to assume such an interpreter would need `16-bit`, but I'm not sure.

So overall, keeping it optional is meant to provide room for exotic interpreters, but it maybe reasonable to tighten it down and rather ask maintainers of such interpreters to come forward with required additions.
I guess that is reason enough to have this discussion with the wider community.